core/GlobalState.h

```diff
@@ -188,6 +188,16 @@ class GlobalState final {
     // Indicates the number of times LSP has run the type checker with this global state.
     // Used to ensure GlobalState is in the correct state to process requests.
     unsigned int lspTypecheckCount = 0;
+    // In LSP mode, contains an integer that increments every time an edit comes in. If lspEpoch changes
+    // mid-typechecking, then new user edits are available to typecheck. May wrap around. Used to implement canceling
+    // long-running typechecking operations.
+    // Defaults to a `nullptr`, which disables this functionality.
+    std::shared_ptr<std::atomic<u4>> lspEpoch;
+    // When a cancelable typechecking operation is active, contains the expected `lspEpoch` value. If it differs from
+    // `lspEpoch`, then typechecking should be canceled.
+    std::optional<u4> expectedLspEpoch = std::nullopt;
```

Comment:
I think replacing this pair of fields with 

```
std::shared_ptr<std::atomic<u4>> currentlyProcessingLSPEpoch;
std::shared_ptr<std::atomic<u4>> lspEpochInvalidator; // should always be `>= currentlyProcessingLSPEpoch`(modulo overflows). if value is different from currentlyProcessingLSPEpoch than LSP wants the current request to be cancelled. `currentlyProcessingLSPEpoch`  is bumped by typechecking thread, `lspEpochInvalidator` is bumped by the thread that decides to perform cancellation.
```
I think it will be both easier to reason about them in docs and tp